### PR TITLE
[RFC] Secure key serivces

### PR DESCRIPTION
**Pull request by etienne-lms**
Related to:
* Client lib from optee_client : https://github.com/OP-TEE/optee_client/pull/113
* Test from optee_xtest : https://github.com/OP-TEE/optee_test/pull/247
* Build: https://github.com/OP-TEE/build/pull/219

### CODE REVIEW

File: ta_services/secure_key_services/include/sks_ta.h [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+#ifndef __SKS_TA_H__
+#define __SKS_TA_H__
+
+#include <sys/types.h>
+#include <stdint.h>
+
+#define TA_SKS_UUID { 0xfd02c9da, 0x306c, 0x48c7, \
+                        { 0xa4, 0x9c, 0xbb, 0xd8, 0x27, 0xae, 0x86, 0xee } }
+
+#define KEYCMD			0
```
Comment (jenswi-linaro) [inline ta_services/secure_key_services/include/sks_ta.h]: There seem to be some structure of how the different bits in the command id are assigned. Please add a comment describing the structure and the purpose.

File: ta_services/secure_key_services/include/sks_ta.h [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	uint8_t hardwareVersion[2];
+	uint8_t firmwareVersion[2];
+	uint8_t utcTime[16];
+};
+
+/*
+ * Get list of the supported mechanisms
+ * request args:	[32b-slot-id]
+ * input data:		none
+ * output data:		[array-of-sks-mechanism_IDs]
```
Comment (jenswi-linaro) [inline ta_services/secure_key_services/include/sks_ta.h]: What is a sks-mechanism_ID?

File: ta_services/secure_key_services/include/sks_ta.h [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+ */
+#define SKS_CMD_CK_MECHANISM_IDS	TA_SKS_CMD_CRYPTOKI(0x000003)
+
+/*
+ * Open Read-Only Session
+ * request args:	[32b-slot-id][32b-mechanism-id]
+ * input data:		none
+ * output data:		[sks-mechanism-info]
```
Comment (jenswi-linaro) [inline ta_services/secure_key_services/include/sks_ta.h]: What is sks-mechanism-info?

File: ta_services/secure_key_services/include/sks_ta.h [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+/*
+ * request args:	none
+ * input data:		none
+ * output data:		none
+ */
+#define SKS_CMD_CK_PING			TA_SKS_CMD_CRYPTOKI(0x000000)
+
+/*
+ * request args:	[32b-slot-id]
```
Comment (jenswi-linaro) [inline ta_services/secure_key_services/include/sks_ta.h]: For all the command ids: please specify exactly how each of the used four parameters are used.

**Comment by jenswi-linaro** [comment]
With the `pkcs11.h`, `pkcs11f.h` and `pkcs11t.h` files as is you add a lot of stuff that never will be used.  3207 lines just for a few defines and types needed in the api descriptions of the TA? This needs to be a stable ABI so we need full control over it, if there's some update of the PKCS#11 standard we can't blindly just copy the new files.

I would prefer a TA API which is built up as features are added. We should provide what's needed to implement a PKCS#11 API in user space, but that doesn't mean that the TA API must use the PKCS#11 headers directly. This would make it easier to review the code and also to get an idea of what's supported.

File: ta_services/secure_key_services/src/entry.c [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+void TA_CloseSessionEntryPoint(void *session)
+{
+	ck_token_close_tee_session((int)session);
+	TEE_Free(handle_put(&session_db, (int)session));
+}
+
+/*
+ * Entry point for SKS TA commands
+ *
+ * ABI: param#0 is ctrl buffer: serialazed arguments
```
Comment (jbech-linaro) [inline ta_services/secure_key_services/src/entry.c]: s/serialazed/serialized/

**Comment by etienne-lms** [comment]
Updated with the TA API proposal.
This is aligned with  OP-TEE/optee_client#113, OP-TEE/optee_test#247 and OP-TEE/build#219.

Still  a draft version. I need to formalize the processing sequences and add other AES alternate processing (generate, derive, ...).

File: ta_services/secure_key_services/src/handle.c [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+	if (!db || handle < 0 || (size_t)handle >= db->max_ptrs)
+		return NULL;
+
+	p = db->ptrs[handle];
+	db->ptrs[handle] = NULL;
+	return p;
+}
+
+void *handle_lookup(struct handle_db *db, int handle)
```
Comment (jenswi-linaro) [inline ta_services/secure_key_services/src/handle.c]: This function and `handle_put()` two examples of functions that are sensitive to Spectre. There's a macro just being developed in the kernel that deals with this kind of data structure to be Spectre-safe.

**Comment by etienne-lms** [comment]
Updated. 
I noted the handle database that is currently weak and should be replace with something safer 2nd commit affected).

Start from **sks_ta.h** in the 1st commit to see the TA API.
Following commits introduce basics for token management.
Then comes the commits for for TA commands and related services.

An ugly part that could be cleaned/simplified are the serializer/serialize functions.

**Comment by etienne-lms** [comment]
Closing this P-R. The feature will be re-submitted once matured.
One can still track on-going developments from https://github.com/etienne-lms/optee_os (branches sks and sks-next).

**Pull request closed without merging**